### PR TITLE
PP-4141 Search for first 6 digits and last 4 digits of card number

### PR DESCRIPTION
**Pull request by simad**
## WHAT
- Added support for searching for first 6 digits and last 4 digits of card number
- Both of those have to be numeric and not empty to be valid. We return a bad request if any of those parameters are invalid. If not specified (ie. `""`) we ignore the query params

### CODE REVIEW

File: src/test/java/uk/gov/pay/api/it/validation/PaymentResourceSearchValidationITest.java [hunk 1]
```diff
@@ -21,6 +21,8 @@
 public class PaymentResourceSearchValidationITest extends PaymentResourceITestBase {
 
     private static final String VALID_REFERENCE = "test_reference";
+    private static final String VALID_LAST_DIGITS_CARD_NUMBER = "4242";
+    private static final String VALID_FIRST_DIGITS_CARD_NUMBER = "424242";
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/api/it/validation/PaymentResourceSearchValidationITest.java]: Minor but I’d feel slightly more confident if one of these strings weren’t a subset of the other…

File: src/test/java/uk/gov/pay/api/it/validation/PaymentResourceSearchValidationITest.java [hunk 2]
```diff
@@ -166,6 +172,31 @@ public void searchPayments_errorWhenFromAndToDatesAreNotInZoneDateTimeFormat() t
...
+    
+    @Test
+    public void searchPayments_errorWhenCardDigitsAreInvalid() throws Exception {
+        InputStream body = searchPayments(API_KEY,
+                ImmutableMap.<String, String>builder()
+                        .put("reference", VALID_REFERENCE)
+                        .put("email", VALID_EMAIL)
+                        .put("state", VALID_STATE)
+                        .put("first_digits_card_number", "4242")
+                        .put("last_digits_card_number", "423")
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/api/it/validation/PaymentResourceSearchValidationITest.java]: Maybe have one of them be the wrong length and the other one be the right length but with non-digits in it to cover a broader swathe of the validation

File: src/test/java/uk/gov/pay/api/validation/PaymentSearchValidatorTest.java [hunk 3]
```diff
@@ -144,6 +144,42 @@ public void validateParams_shouldGiveAnError_forInvalidDirectDebitState() {
...
+                "","", "", "", "422");
+    }
+
+    @Test
+    public void validateParams_shouldGiveAnError_forInvalidLastDigits() {
+        expectedException.expect(aValidationExceptionContaining("P0401", "Invalid parameters: last_digits_card_number. See Public API documentation for the correct data formats"));
+        PaymentSearchValidator.validateSearchParameters(account,
+                "", "", "",
+                "", "", "",
+                "","", "", "", "422a");
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/api/validation/PaymentSearchValidatorTest.java]: Negative numbers and non-Arabic digits would make other good tests

File: src/main/java/uk/gov/pay/api/service/PaymentSearchService.java [hunk 4]
```diff
@@ -17,15 +17,18 @@
...
-    private static final String DISPLAY_SIZE = "display_size";
-    private static final String AGREEMENT_KEY = "agreement_id";
+    public static final String REFERENCE_KEY = "reference";
+    public static final String EMAIL_KEY = "email";
+    public static final String STATE_KEY = "state";
+    public static final String CARD_BRAND_KEY = "card_brand";
+    public static final String FROM_DATE_KEY = "from_date";
+    public static final String TO_DATE_KEY = "to_date";
+    public static final String PAGE = "page";
+    public static final String DISPLAY_SIZE = "display_size";
```
[the remark this answers is not in the document]
Comment (simad) [reply src/main/java/uk/gov/pay/api/service/PaymentSearchService.java]: yes I did that originally but it broke end to end. We can edit that but it's a bit cumbersome

File: src/main/java/uk/gov/pay/api/service/PaymentSearchService.java [hunk 5]
```diff
@@ -59,7 +62,10 @@ public Response doSearch(Account account, String reference, String email, String
         queryParams.put(EMAIL_KEY, email);
         queryParams.put(STATE_KEY, state);
         queryParams.put(CARD_BRAND_KEY, cardBrand);
+        queryParams.put(CARDHOLDER_NAME_KEY, cardHolderName);
```
[the remark this answers is not in the document]
Comment (simad) [reply src/main/java/uk/gov/pay/api/service/PaymentSearchService.java]: had a chat with Katie about this

File: src/main/java/uk/gov/pay/api/validation/PaymentSearchValidator.java [hunk 6]
```diff
@@ -55,7 +69,19 @@ public static void validateSearchParameters(Account account, String state, Strin
 
     private static void validateAgreement(String agreement, List<String> validationErrors) {
         if (!isValid(agreement, AGREEMENT_ID_MAX_LENGTH)){
-                validationErrors.add("agreement");
+            validationErrors.add("agreement");
+        }
+    }
+
+    private static void validatFirstDigitsCardNumber(String firstDigitsCardNumber, List<String> validationErrors) {
```
[the remark this answers is not in the document]
Comment (simad) [reply src/main/java/uk/gov/pay/api/validation/PaymentSearchValidator.java]: ops

File: src/main/java/uk/gov/pay/api/validation/ExactLengthValidator.java [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package uk.gov.pay.api.validation;
+
+import static org.eclipse.jetty.util.StringUtil.isBlank;
```
[the remark this answers is not in the document]
Comment (simad) [reply src/main/java/uk/gov/pay/api/validation/ExactLengthValidator.java]: no :D

File: src/main/java/uk/gov/pay/api/validation/ExactLengthValidator.java [hunk 8]
```diff
@@ -0,0 +1,9 @@
...
+
+class ExactLengthValidator {
+    static boolean isValid(String value, int length) {
+        return isBlank(value) || value.length() == length;
```
[the remark this answers is not in the document]
Comment (simad) [reply src/main/java/uk/gov/pay/api/validation/ExactLengthValidator.java]: we have a isBlank check for all of those validators (ie. MaxLength and so on) I didn't want to change that, as I would have to modify all of the validators to make it consistent

File: src/main/java/uk/gov/pay/api/validation/NumericValidator.java [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package uk.gov.pay.api.validation;
+
+import static org.apache.commons.lang3.StringUtils.isNumeric;
+import static org.eclipse.jetty.util.StringUtil.isBlank;
+
+class NumericValidator {
+    static boolean isValid(String value) {
+        return isBlank(value) || isNumeric(value);
```
[the remark this answers is not in the document]
Comment (simad) [reply src/main/java/uk/gov/pay/api/validation/NumericValidator.java]: dafuq

[the remark this answers is not in the document]
Comment (simad) [reply src/main/java/uk/gov/pay/api/validation/NumericValidator.java]: are you being racist towards people speaking Devanagari? This is outrageous

File: src/main/java/uk/gov/pay/api/validation/PaymentSearchValidator.java [hunk 10]
```diff
@@ -55,7 +69,19 @@ public static void validateSearchParameters(Account account, String state, Strin
 
     private static void validateAgreement(String agreement, List<String> validationErrors) {
         if (!isValid(agreement, AGREEMENT_ID_MAX_LENGTH)){
-                validationErrors.add("agreement");
+            validationErrors.add("agreement");
```
[the remark this answers is not in the document]
Comment (simad) [reply src/main/java/uk/gov/pay/api/validation/PaymentSearchValidator.java]: yes

File: src/main/java/uk/gov/pay/api/validation/PaymentSearchValidator.java [hunk 11]
```diff
@@ -55,7 +69,19 @@ public static void validateSearchParameters(Account account, String state, Strin
...
+        }
+    }
+
+    private static void validatFirstDigitsCardNumber(String firstDigitsCardNumber, List<String> validationErrors) {
+        if (!(ExactLengthValidator.isValid(firstDigitsCardNumber, FIRST_DIGITS_CARD_NUMBER_LENGTH) && NumericValidator.isValid(firstDigitsCardNumber))) {
```
[the remark this answers is not in the document]
Comment (simad) [reply src/main/java/uk/gov/pay/api/validation/PaymentSearchValidator.java]: 👍

File: src/test/java/uk/gov/pay/api/it/PaymentResourceSearchITest.java [hunk 12]
```diff
@@ -248,6 +252,76 @@ public void searchPayments_filterByEmail() {
...
     }
 
+    @Test
+    public void searchPayments_filterByLastDigitsCardNumber() {
+        String payments = aPaginatedPaymentSearchResult()
+                .withCount(10)
+                .withPage(2)
+                .withTotal(20)
+                .withPayments(aSuccessfulSearchPayment()
+                        .withMatchingCardDetails(CARD_DETAILS)
```
[the remark this answers is not in the document]
Comment (simad) [reply src/test/java/uk/gov/pay/api/it/PaymentResourceSearchITest.java]: oof

File: src/test/java/uk/gov/pay/api/it/validation/PaymentResourceSearchValidationITest.java (hunk 1 above)
Comment (simad) [reply src/test/java/uk/gov/pay/api/it/validation/PaymentResourceSearchValidationITest.java]: 👍

File: src/test/java/uk/gov/pay/api/it/validation/PaymentResourceSearchValidationITest.java (hunk 2 above)
Comment (simad) [reply src/test/java/uk/gov/pay/api/it/validation/PaymentResourceSearchValidationITest.java]: 👍

File: src/main/java/uk/gov/pay/api/service/PaymentSearchService.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (alexbishop1) [reply src/main/java/uk/gov/pay/api/service/PaymentSearchService.java]: Reordering the constants broke end-to-end tests?

File: src/test/java/uk/gov/pay/api/it/PaymentResourceSearchITest.java [hunk 13]
```diff
@@ -495,7 +569,7 @@ public void searchPayments_withMandateId_whenCardPayment_shouldReturnABadRequest
         JsonAssert.with(body)
                 .assertThat("$.*", hasSize(2))
                 .assertThat("$.code", is("P0401"))
-                .assertThat("$.description", is("Invalid parameters: agreement_id. See Public API documentation for the correct data formats"));
+                .assertThat("$.description", is("Invalid parameters: agreement_id=my_agreement. See Public API documentation for the correct data formats"));
```
[the remark this answers is not in the document]
Comment (alexbishop1) [reply src/test/java/uk/gov/pay/api/it/PaymentResourceSearchITest.java]: After discussion elsewhere, this has been changed back

**Pull request merged**